cf-custom-resources/lib/custom-domain-app-runner.js

```diff
@@ -234,25 +200,134 @@ async function validateCertForDomain(serviceARN, domainName) {
         });
 
         lastDomainStatus = domain.Status;
-        if (lastDomainStatus !== DOMAIN_STATUS_PENDING_VERIFICATION) {
+
+        if (!domainValidationRecordReady(domain)) {
             await sleep(3000);
             continue;
         }
+
         // Upsert all records needed for certificate validation.
         const records = domain.CertificateValidationRecords;
+        let promises = [];
         for (const record of records) {
-            await updateCNAMERecordAndWait(record.Name, record.Value, appHostedZoneID, "UPSERT").catch(err => {
-                throw new Error(`update validation records for domain ${domainName}: ` + err.message);
-            });
+            promises.push(
+                updateCNAMERecordAndWait(record.Name, record.Value, appHostedZoneID, "UPSERT").catch(err => {
+                    throw new Error(`update validation records for domain ${domainName}: ` + err.message);
+                })
+            );
         }
-        break;
+        return Promise.all(promises);
     }
 
     if (i === ATTEMPTS_WAIT_FOR_PENDING) {
         throw new Error(`update validation records for domain ${domainName}: fail to wait for state ${DOMAIN_STATUS_PENDING_VERIFICATION}, stuck in ${lastDomainStatus}`);
     }
 }
 
+/**
+ * There are one known scenarios where status could be ACTIVE right after it's associated:
+ * When the domain just got deleted and added again. In this case, even though the validation records could
+ * have been deleted, the previously successful validation results are still cached. Because of the cache,
+ * the domain will show to be ACTIVE immediately after it's associated , although the validation records are not
+ * there anymore.
+ * In this case, the status won't transit to PENDING_VERIFICATION, so we need to check whether the validation
+ * records are ready by counting if there are three of them.
+ *
+ * @param {string} domain
+ * @returns {boolean}
+ */
+function domainValidationRecordReady(domain) {
+    if (domain.Status === DOMAIN_STATUS_PENDING_VERIFICATION) {
+        return true;
+    }
+
+    if (domain.Status === DOMAIN_STATUS_ACTIVE && domain.CertificateValidationRecords && domain.CertificateValidationRecords.length === 3) {
```

Comment:
I don't think that I understand why we need to check if there are 3 of them? What's the prob with returning true if the status is just active?

Reply:
Because we want to make sure all 3 of the validation records are ready to be upserted. I'm not sure if it's safe to make the assumption that if the status is active then all records are ready.